Static local function should not be able to call non-static generic local function from parent scope

**Version Used**: VS2019 16.4.0 Preview3

Seems fix for #38129 not works for generic local function.

**Steps to Reproduce**:

https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMBGAsAKFygZgAJ0iBhAG2igyIG9cinmmUMA2EgFiIFk4YACwD2AEwAUASkYtmDHLMVEAlgDsYKogF4iABgDcMpcxQ8AMsIDGECgB4AKgD4pR4/Vdvmy7XsMLPRAC+HkpsaEQQPlLajioA1HF+ASQYAJziEJJJAWycpkQAyjAQMMqWFtYULv6e8sksFTa2ajDOWSGKwTVuRSVljVXt3SxdgUA===
```cs
        static void Method()
        {
            int i = 0;
            void Local<T>()
            {
                i = 0;
            }
            Action a = () => i++;
            Console.WriteLine(a);
            static void StaticLocal()
            {
                Local<int>();
            }
            StaticLocal();
        }
```

**Expected Behavior**:
It should not compile. And get an error as it did when used without generic.
`error CS8421: A static local function cannot contain a reference to 'Local'.`
See `Item 2`(#38129) in https://github.com/dotnet/roslyn/blob/master/docs/compilers/CSharp/Compiler%20Breaking%20Changes%20-%20post%20VS2019.md

**Actual Behavior**:
It compile without error, and get an non-static method.

--- a/src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_ExpressionTrees.cs
+++ b/src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_ExpressionTrees.cs
@@ -147,7 +147,7 @@ namespace Microsoft.CodeAnalysis.CSharp
 
         private void CheckReferenceToMethodIfLocalFunction(BoundExpression node, MethodSymbol method)
         {
-            if (method.OriginalDefinition is LocalFunctionSymbol localFunction)
+            if (method?.OriginalDefinition is LocalFunctionSymbol localFunction)
             {
                 CheckReferenceToVariable(node, localFunction);
             }
@@ -664,11 +664,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             }
 
             CheckReceiverIfField(node.ReceiverOpt);
-
-            if (!(method is null))
-            {
-                CheckReferenceToMethodIfLocalFunction(node, method);
-            }
+            CheckReferenceToMethodIfLocalFunction(node, method);
 
             if (method is null || method.RequiresInstanceReceiver)
             {

--- a/src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_ExpressionTrees.cs
+++ b/src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_ExpressionTrees.cs
@@ -145,6 +145,13 @@ namespace Microsoft.CodeAnalysis.CSharp
             }
         }
 
+        private void CheckReferenceToMethodIfLocalFunction(BoundExpression node, MethodSymbol method)
+        {
+            if (method.OriginalDefinition is LocalFunctionSymbol localFunction)
+            {
+                CheckReferenceToVariable(node, localFunction);
+            }
+        }
 
         public override BoundNode VisitConvertedSwitchExpression(BoundConvertedSwitchExpression node)
         {
@@ -349,10 +356,7 @@ namespace Microsoft.CodeAnalysis.CSharp
         {
             VisitCall(node.Method, null, node.Arguments, node.ArgumentRefKindsOpt, node.ArgumentNamesOpt, node.Expanded, node);
             CheckReceiverIfField(node.ReceiverOpt);
-            if (node.Method is LocalFunctionSymbol)
-            {
-                CheckReferenceToVariable(node, node.Method);
-            }
+            CheckReferenceToMethodIfLocalFunction(node, node.Method);
             return base.VisitCall(node);
         }
 
@@ -661,12 +665,12 @@ namespace Microsoft.CodeAnalysis.CSharp
 
             CheckReceiverIfField(node.ReceiverOpt);
 
-            if (method is LocalFunctionSymbol)
+            if (!(method is null))
             {
-                CheckReferenceToVariable(node, method);
+                CheckReferenceToMethodIfLocalFunction(node, method);
             }
 
-            if (method == null || method.RequiresInstanceReceiver)
+            if (method is null || method.RequiresInstanceReceiver)
             {
                 Visit(node.ReceiverOpt);
             }

--- a/src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs
+++ b/src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs
@@ -462,8 +462,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                     topLevelMethodId,
                     originalMethod,
                     closure.BlockSyntax,
-                    lambdaId,
-                    Diagnostics);
+                    lambdaId);
                 closure.SynthesizedLoweredMethod = synthesizedMethod;
             });
 

--- a/src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/SynthesizedClosureMethod.cs
+++ b/src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/SynthesizedClosureMethod.cs
@@ -30,8 +30,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             DebugId topLevelMethodId,
             MethodSymbol originalMethod,
             SyntaxReference blockSyntax,
-            DebugId lambdaId,
-            DiagnosticBag diagnostics)
+            DebugId lambdaId)
             : base(containingType,
                    originalMethod,
                    blockSyntax,
@@ -101,6 +100,9 @@ namespace Microsoft.CodeAnalysis.CSharp
             }
 
             AssignTypeMapAndTypeParameters(typeMap, typeParameters);
+
+            // static local functions should be emitted as static.
+            Debug.Assert(!(originalMethod is LocalFunctionSymbol) || !originalMethod.IsStatic || IsStatic);
         }
 
         private static DeclarationModifiers MakeDeclarationModifiers(ClosureKind closureKind, MethodSymbol originalMethod)

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs
@@ -4922,33 +4922,52 @@ class C
         public void StaticLocalFunctionLocalFunctionReference_02()
         {
             var source =
-@"#pragma warning disable 0219
-#pragma warning disable 8321
+@"#pragma warning disable 8321
 class Program
 {
     static void Method()
     {
-        int i = 0;
-        void Local1<T>()
-        {
-            i = 0;
-        }
-        void Local2<T>() { }
+        void Local<T>() {}
         static void StaticLocal()
         {
-            Local1<int>();
-            Local2<int>();
+            Local<int>();
         }
     }
 }";
             var comp = CreateCompilation(source);
             comp.VerifyDiagnostics(
-                // (15,13): error CS8421: A static local function cannot contain a reference to 'Local1'.
-                //             Local1<int>();
-                Diagnostic(ErrorCode.ERR_StaticLocalFunctionCannotCaptureVariable, "Local1<int>()").WithArguments("Local1").WithLocation(15, 13),
-                // (16,13): error CS8421: A static local function cannot contain a reference to 'Local2'.
-                //             Local2<int>();
-                Diagnostic(ErrorCode.ERR_StaticLocalFunctionCannotCaptureVariable, "Local2<int>()").WithArguments("Local2").WithLocation(16, 13));
+                // (9,13): error CS8421: A static local function cannot contain a reference to 'Local'.
+                //             Local<int>();
+                Diagnostic(ErrorCode.ERR_StaticLocalFunctionCannotCaptureVariable, "Local<int>()").WithArguments("Local").WithLocation(9, 13));
+        }
+
+        [Fact, WorkItem(39706, "https://github.com/dotnet/roslyn/issues/39706")]
+        public void StaticLocalFunctionLocalFunctionReference_03()
+        {
+            var source =
+@"using System;
+class Program
+{
+    static void Method()
+    {
+        int i = 0;
+        void Local<T>()
+        {
+            i = 0;
+        }
+        Action a = () => i++;
+        static void StaticLocal()
+        {
+            Local<int>();
+        }
+        StaticLocal();
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (14,13): error CS8421: A static local function cannot contain a reference to 'Local'.
+                //             Local<int>();
+                Diagnostic(ErrorCode.ERR_StaticLocalFunctionCannotCaptureVariable, "Local<int>()").WithArguments("Local").WithLocation(14, 13));
         }
 
         [Fact, WorkItem(38240, "https://github.com/dotnet/roslyn/issues/38240")]
@@ -4989,7 +5008,7 @@ class Program
 {
     static void Method()
     {
-        void Local<T>() { }
+        void Local<T>() {}
         static System.Action StaticLocal()
         {
             return Local<int>;
@@ -5001,6 +5020,62 @@ class Program
                 // (9,20): error CS8421: A static local function cannot contain a reference to 'Local'.
                 //             return Local<int>;
                 Diagnostic(ErrorCode.ERR_StaticLocalFunctionCannotCaptureVariable, "Local<int>").WithArguments("Local").WithLocation(9, 20));
+        }
+
+        [Fact, WorkItem(39706, "https://github.com/dotnet/roslyn/issues/39706")]
+        public void StaticLocalFunctionLocalFunctionDelegateReference_03()
+        {
+            var source =
+@"using System;
+class Program
+{
+    static void Method()
+    {
+        int i = 0;
+        void Local<T>()
+        {
+            i = 0;
+        }
+        Action a = () => i++;
+        a = StaticLocal();
+        static Action StaticLocal()
+        {
+            return Local<int>;
+        }
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (15,20): error CS8421: A static local function cannot contain a reference to 'Local'.
+                //             return Local<int>;
+                Diagnostic(ErrorCode.ERR_StaticLocalFunctionCannotCaptureVariable, "Local<int>").WithArguments("Local").WithLocation(15, 20));
+        }
+
+        [Fact]
+        public void StaticLocalFunctionGenericStaticLocalFunction()
+        {
+            var source =
+@"using System;
+class Program
+{
+    static void Main()
+    {
+        static void F1<T>()
+        {
+            Console.WriteLine(typeof(T));
+        }
+        static void F2()
+        {
+            F1<int>();
+            Action a = F1<string>;
+            a();
+        }
+        F2();
+    }
+}";
+            CompileAndVerify(source, expectedOutput:
+@"System.Int32
+System.String");
         }
 
         [Fact, WorkItem(38240, "https://github.com/dotnet/roslyn/issues/38240")]

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs
@@ -4893,7 +4893,7 @@ class Program
         }
 
         [Fact, WorkItem(38129, "https://github.com/dotnet/roslyn/issues/38129")]
-        public void StaticLocalFunctionLocalFunctionReference()
+        public void StaticLocalFunctionLocalFunctionReference_01()
         {
             var source =
 @"#pragma warning disable 8321
@@ -4918,8 +4918,41 @@ class C
                 Diagnostic(ErrorCode.ERR_StaticLocalFunctionCannotCaptureVariable, "F1()").WithArguments("F1").WithLocation(11, 13));
         }
 
+        [Fact, WorkItem(39706, "https://github.com/dotnet/roslyn/issues/39706")]
+        public void StaticLocalFunctionLocalFunctionReference_02()
+        {
+            var source =
+@"#pragma warning disable 0219
+#pragma warning disable 8321
+class Program
+{
+    static void Method()
+    {
+        int i = 0;
+        void Local1<T>()
+        {
+            i = 0;
+        }
+        void Local2<T>() { }
+        static void StaticLocal()
+        {
+            Local1<int>();
+            Local2<int>();
+        }
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (15,13): error CS8421: A static local function cannot contain a reference to 'Local1'.
+                //             Local1<int>();
+                Diagnostic(ErrorCode.ERR_StaticLocalFunctionCannotCaptureVariable, "Local1<int>()").WithArguments("Local1").WithLocation(15, 13),
+                // (16,13): error CS8421: A static local function cannot contain a reference to 'Local2'.
+                //             Local2<int>();
+                Diagnostic(ErrorCode.ERR_StaticLocalFunctionCannotCaptureVariable, "Local2<int>()").WithArguments("Local2").WithLocation(16, 13));
+        }
+
         [Fact, WorkItem(38240, "https://github.com/dotnet/roslyn/issues/38240")]
-        public void StaticLocalFunctionLocalFunctionDelegateReference()
+        public void StaticLocalFunctionLocalFunctionDelegateReference_01()
         {
             var source =
 @"#pragma warning disable 8321
@@ -4945,6 +4978,29 @@ class C
                 // (12,28): error CS8421: A static local function cannot contain a reference to 'F1'.
                 //             _ = new Action(F1);
                 Diagnostic(ErrorCode.ERR_StaticLocalFunctionCannotCaptureVariable, "F1").WithArguments("F1").WithLocation(12, 28));
+        }
+
+        [Fact, WorkItem(39706, "https://github.com/dotnet/roslyn/issues/39706")]
+        public void StaticLocalFunctionLocalFunctionDelegateReference_02()
+        {
+            var source =
+@"#pragma warning disable 8321
+class Program
+{
+    static void Method()
+    {
+        void Local<T>() { }
+        static System.Action StaticLocal()
+        {
+            return Local<int>;
+        }
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (9,20): error CS8421: A static local function cannot contain a reference to 'Local'.
+                //             return Local<int>;
+                Diagnostic(ErrorCode.ERR_StaticLocalFunctionCannotCaptureVariable, "Local<int>").WithArguments("Local").WithLocation(9, 20));
         }
 
         [Fact, WorkItem(38240, "https://github.com/dotnet/roslyn/issues/38240")]

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs
@@ -5004,22 +5004,28 @@ class C
         {
             var source =
 @"#pragma warning disable 8321
+using System;
 class Program
 {
     static void Method()
     {
         void Local<T>() {}
-        static System.Action StaticLocal()
+        static void StaticLocal()
         {
-            return Local<int>;
+            Action a;
+            a = Local<int>;
+            a = new Action(Local<string>);
         }
     }
 }";
             var comp = CreateCompilation(source);
             comp.VerifyDiagnostics(
-                // (9,20): error CS8421: A static local function cannot contain a reference to 'Local'.
-                //             return Local<int>;
-                Diagnostic(ErrorCode.ERR_StaticLocalFunctionCannotCaptureVariable, "Local<int>").WithArguments("Local").WithLocation(9, 20));
+                // (11,17): error CS8421: A static local function cannot contain a reference to 'Local'.
+                //             a = Local<int>;
+                Diagnostic(ErrorCode.ERR_StaticLocalFunctionCannotCaptureVariable, "Local<int>").WithArguments("Local").WithLocation(11, 17),
+                // (12,28): error CS8421: A static local function cannot contain a reference to 'Local'.
+                //             a = new Action(Local<string>);
+                Diagnostic(ErrorCode.ERR_StaticLocalFunctionCannotCaptureVariable, "Local<string>").WithArguments("Local").WithLocation(12, 28));
         }
 
         [Fact, WorkItem(39706, "https://github.com/dotnet/roslyn/issues/39706")]

